File: Category/01_Data Structure/SegTree/SegTree.cpp
template <class Info, class Tag>
class SegTree {
private:
    int n;
    vector<Info> info;
    vector<Tag> tag;

    void apply(int o, const Tag& t) { tag[o].apply(t), info[o].apply(t); }

    void push(int o) { apply(o << 1, tag[o]), apply(o << 1 | 1, tag[o]), tag[o] = Tag(); }

    void pull(int o) { info[o] = info[o << 1] + info[o << 1 | 1]; }

public:
    SegTree() : n(0) {}
    SegTree(int n, Info v = Info {}) { init(vector(n, v)); }
    SegTree(const vector<Info>& _init) { init(_init); }

    void init(const vector<Info>& _init) {
        n = _init.size();
        info.assign(4 << __lg(n), Info {});
        tag.assign(4 << __lg(n), Tag());

        auto build = [&](auto&& self, int o, int l, int r) -> void {
            if (l == r) {
                info[o] = _init[l];
                return;
            }
            int m = l + (r - l) / 2;
            self(self, o << 1, l, m);
            self(self, o << 1 | 1, m + 1, r);
            pull(o);
        };
        build(build, 1, 0, n - 1);
    }

    void modify(int p, const Info& _init, int o, int l, int r) {
        assert(0 <= p && p < n);
        if (l == r) {
            info[o] = _init;
            return;
        }
        push(o);
        int m = l + (r - l) / 2;
        if (p <= m) {
            modify(p, _init, o << 1, l, m);
        } else {
            modify(p, _init, o << 1 | 1, m + 1, r);
        }
        pull(o);
    }
    void modify(int p, const Info& _init) { modify(p, _init, 1, 0, n - 1); }

    void modify(int L, int R, const Tag& tag, int o, int l, int r) {
        assert(0 <= L && L <= R && R < n);
        if (L <= l && r <= R) {
            apply(o, tag);
            return;
        }
        push(o);
        int m = l + (r - l) / 2;
        if (L <= m) {
            modify(L, R, tag, o << 1, l, m);
        }
        if (m < R) {
            modify(L, R, tag, o << 1 | 1, m + 1, r);
        }
        pull(o);
    }
    void modify(int L, int R, const Tag& tag) { modify(L, R, tag, 1, 0, n - 1); }

    Info query(int p, int o, int l, int r) {
        assert(0 <= p && p < n);
        if (l == r) {
            return info[o];
        }
        push(o);
        int m = l + (r - l) / 2;
        if (p <= m) {
            return query(p, o << 1, l, m);
        }
        return query(p, o << 1 | 1, m + 1, r);
    }
    Info query(int p) { return query(p, 1, 0, n - 1); }

    Info query(int L, int R, int o, int l, int r) {
        assert(0 <= L && L <= R && R < n);
        if (L <= l && r <= R) {
            return info[o];
        }
        push(o);
        int m = l + (r - l) / 2;
        if (R <= m) {
            return query(L, R, o << 1, l, m);
        }
        if (m < L) {
            return query(L, R, o << 1 | 1, m + 1, r);
        }
        return query(L, R, o << 1, l, m) + query(L, R, o << 1 | 1, m + 1, r);
    }
    Info query(int L, int R) { return query(L, R, 1, 0, n - 1); }

    Info queryAll() { return info[1]; }

    int findFirst(int L, int R, auto&& pred, int o, int l, int r) {
        assert(0 <= L && L <= R && R < n);
        if (l > R || r < L || !pred(info[o])) {
            return -1;
        }
        if (l == r) {
            return l;
        }
        int m = l + (r - l) / 2;
        push(o);
        int idx = findFirst(L, R, pred, o << 1, l, m);
        if (idx == -1) {
            idx = findFirst(L, R, pred, o << 1 | 1, m + 1, r);
        }
        return idx;
    }
    int findFirst(int l, int r, auto&& pred) { return findFirst(l, r, pred, 1, 0, n - 1); }

    int findLast(int L, int R, auto&& pred, int o, int l, int r) {
        assert(0 <= L && L <= R && R < n);
        if (l > R || r < L || !pred(info[o])) {
            return -1;
        }
        if (l == r) {
            return l;
        }
        int m = l + (r - l) / 2;
        push(o);
        int idx = findLast(L, R, pred, o << 1 | 1, m + 1, r);
        if (idx == -1) {
            idx = findLast(L, R, pred, o << 1, l, m);
        }
        return idx;
    }
    int findLast(int l, int r, auto&& pred) { return findLast(l, r, pred, 1, 0, n - 1); }
};

struct Tag {
    i64 add = 0;
    Tag() {}
    Tag(i64 v) : add(v) {}
    void apply(const Tag& t) {
        if (t.add) {
            add += t.add;
        }
    }
};
struct Info {
    i64 sum = 0, len = 1;
    void apply(const Tag& t) {
        if (t.add) {
            sum += len * t.add;
        }
    }
};
Info operator+(const Info& a, const Info& b) {
    Info info;
    info.sum = a.sum + b.sum;
    info.len = a.len + b.len;
    return info;
}

File: Go/SegTree.go
type SegTree struct {
    n    int
    info []Info
    tag  []Tag
}

func NewSegTree(init []Info) *SegTree {
    n := len(init)
    if n == 0 {
        panic("SegTree: init length must be > 0")
    }
    siz := 2 << bits.Len(uint(n-1))
    seg := &SegTree{
        n:    n,
        info: make([]Info, siz),
        tag:  make([]Tag, siz),
    }
    var build func(o, l, r int)
    build = func(o, l, r int) {
        if l == r {
            seg.info[o] = init[l]
            return
        }
        mid := (l + r) >> 1
        build(o<<1, l, mid)
        build(o<<1|1, mid+1, r)
        seg.info[o] = mergeInfo(seg.info[o<<1], seg.info[o<<1|1])
    }
    build(1, 0, n-1)
    return seg
}

func (seg *SegTree) apply(o int, t Tag) {
    seg.info[o].applyInfo(t)
    seg.tag[o].applyTag(t)
}

func (seg *SegTree) push(o int) {
    seg.apply(o<<1, seg.tag[o])
    seg.apply(o<<1|1, seg.tag[o])
    seg.tag[o] = Tag{}
}

func (seg *SegTree) pull(o int) {
    seg.info[o] = mergeInfo(seg.info[o<<1], seg.info[o<<1|1])
}

func (seg *SegTree) ModifyPoint(p int, v Info) {
    if p < 0 || p >= seg.n {
        panic("SegTree.ModifyPoint: index out of range")
    }
    var dfs func(o, l, r int)
    dfs = func(o, l, r int) {
        if l == r {
            seg.info[o] = v
            return
        }
        seg.push(o)
        mid := (l + r) >> 1
        if p <= mid {
            dfs(o<<1, l, mid)
        } else {
            dfs(o<<1|1, mid+1, r)
        }
        seg.pull(o)
    }
    dfs(1, 0, seg.n-1)
}

func (seg *SegTree) ModifyRange(L, R int, t Tag) {
    if L < 0 || R >= seg.n || L > R {
        panic("SegTree.ModifyRange: invalid range")
    }
    var dfs func(o, l, r int)
    dfs = func(o, l, r int) {
        if L <= l && r <= R {
            seg.apply(o, t)
            return
        }
        seg.push(o)
        mid := (l + r) >> 1
        if L <= mid {
            dfs(o<<1, l, mid)
        }
        if mid < R {
            dfs(o<<1|1, mid+1, r)
        }
        seg.pull(o)
    }
    dfs(1, 0, seg.n-1)
}

func (seg *SegTree) QueryPoint(p int) Info {
    if p < 0 || p >= seg.n {
        panic("SegTree.QueryPoint: index out of range")
    }
    var dfs func(o, l, r int) Info
    dfs = func(o, l, r int) Info {
        if l == r {
            return seg.info[o]
        }
        seg.push(o)
        mid := (l + r) >> 1
        if p <= mid {
            return dfs(o<<1, l, mid)
        }
        return dfs(o<<1|1, mid+1, r)
    }
    return dfs(1, 0, seg.n-1)
}

func (seg *SegTree) QueryRange(L, R int) Info {
    if L < 0 || R >= seg.n || L > R {
        panic("SegTree.QueryRange: invalid range")
    }
    var dfs func(o, l, r int) Info
    dfs = func(o, l, r int) Info {
        if L <= l && r <= R {
            return seg.info[o]
        }
        seg.push(o)
        mid := (l + r) >> 1
        if R <= mid {
            return dfs(o<<1, l, mid)
        }
        if mid < L {
            return dfs(o<<1|1, mid+1, r)
        }
        return mergeInfo(dfs(o<<1, l, mid), dfs(o<<1|1, mid+1, r))
    }
    return dfs(1, 0, seg.n-1)
}

func (seg *SegTree) FindFirst(L, R int, pred func(Info) bool) int {
    if L < 0 || R >= seg.n || L > R {
        return -1
    }
    var dfs func(o, l, r int) int
    dfs = func(o, l, r int) int {
        if r < L || l > R || !pred(seg.info[o]) {
            return -1
        }
        if l == r {
            return l
        }
        seg.push(o)
        mid := (l + r) >> 1
        if j := dfs(o<<1, l, mid); j != -1 {
            return j
        }
        return dfs(o<<1|1, mid+1, r)
    }
    return dfs(1, 0, seg.n-1)
}

func (seg *SegTree) FindLast(L, R int, pred func(Info) bool) int {
    if L < 0 || R >= seg.n || L > R {
        return -1
    }
    var dfs func(o, l, r int) int
    dfs = func(o, l, r int) int {
        if r < L || l > R || !pred(seg.info[o]) {
            return -1
        }
        if l == r {
            return l
        }
        seg.push(o)
        mid := (l + r) >> 1
        if j := dfs(o<<1|1, mid+1, r); j != -1 {
            return j
        }
        return dfs(o<<1, l, mid)
    }
    return dfs(1, 0, seg.n-1)
}

type Tag struct {
    add int
}

func (t *Tag) applyTag(other Tag) {
    if other.add != 0 {
        t.add += other.add
    }
}

type Info struct {
    sum, len int
}

func (info *Info) applyInfo(t Tag) {
    info.sum += info.len * t.add
}

func mergeInfo(a, b Info) Info {
    return Info{
        sum: a.sum + b.sum,
        len: a.len + b.len,
    }
}
